design/debug-collector.md

```diff
@@ -0,0 +1,68 @@
+# Summary
+
+The ekotrace-debug-collector collects and outputs tracer log reports over debug interfaces on a user-defined interval. The user inputs memory locations of tracers or tracer symbols along with the corresponding elf file, and the debug-collector logs reports from those tracers to an output file. The debug collector also allows the user to choose the method of connecting to the target device, either using probe-rs to directly attach, or by connecting to an attached gdb server.
+
+# Motivation
+
+The user may wish to collect ekotrace logs from a microcontroller with no networking capabilities. This tool allows for collection using JTAG/SWD wire protocols instead of UDP.
+
+# Technical Strategy
+
+## Memory Access
+
+To access memory correctly, the program needs to know the word size and endianness of the target. The user can either pass an ELF file containing this information into the CLI: `--elf`, or they can specify the information explicitly: `--word-size 32 --big-endian`.
+
+### Choice of backend
+
+The collector will access memory on the target device either by using the probe-rs Rust library, or by talking to a gdb server that is attached to the target. This choice will be given to the user of the CLI, using `--attach` for probe-rs, or `--gdb-host` to use gdb.
+
+#### probe-rs
+
+probe-rs provides functions to attach to chips over a usb connection, as well as to read from or write to memory locations.
+
+To use this option, users must specify the type of chip that they wish to connect to, which is required by probe-rs: `--attach stm32`.
+
+#### gdb
+
+If the user already has a GDB server that attaches to the chip set up (e.g. OpenOCD), or would rather use such a tool to attach to the target, then they can specify the address of the gdb server that they are running: `--gbd-host localhost:3333`.
+
+In this case, the debug-collector will communicate to the given gdb server over TCP in order to read and write to memory.
+
+### repr(C)
+
+Rust's ABI is undefined, so the representation of structures in memory is unpredictable across compiler versions and architectures. To allow the debug-collector to understand the raw binary it reads from the chip, the ekotrace log data structures must have the `repr(C)` attribute, which tells the compiler to use C's representation of structs in binary. To find the fields of `repr(C)` structs, the debug-collector uses constant offsets for each given word size. This method is fragile to implementation changes to the structs that the debug-collector is accessing.
+
+We would have to add this attribute to `DynamicHistory` and `ChunkedReportState` in ekotrace as well as to `FixedSliceVec` in fixed-slice-vec.
+
+## Symbol resolution
```

Comment:
This will likely require the user to put `no_mangle` attributes on their data; are there any Rust tools/libraries out there that we can use to help alleviate this requirement?